File: rc4.py
import sys
def KSA(key):
    k = list(range(256))
    j = 0
    for i in range(256):
        j = (j + k[i] + key[i % len(key)]) % 256
        k[i], k[j] = k[j], k[i]
    return k

def PRGA(k):
    i = 0
    j = 0
    while True:
        i = (i + 1) % 256
        j = (j + k[i]) % 256
        k[i], k[j] = k[j], k[i]
        K=k[(k[i] + k[j]) % 256]
        yield K

def RC4(k,text):
    stream = PRGA(k)
    for char in text:
        cipher_b= ord(char) ^ next(stream)
        sys.stdout.write("%02X" % (cipher_b))
    print()

if __name__ == '__main__':
    key = 'Key'#input("Key: ")
    text = 'Plaintext'#input("Text: ")
    key = [ord(char) for char in key]
    k = KSA(key)
    RC4(k,text)

File: rc4.cpp
#include <iostream>

using namespace std;

int T[256];// un vector temporar pentru generarea streamului
int S[256];//stream-ul pe care il vom folosi
static string key = "key";//cheia pe care o vom folosi
static string text = "Plaintext";//Textul de encriptat
string enc;
int KSA(){  //Key-Scheduling Algorithm
    for(int i = 0; i<255; i++){
        T[i]=i;
    }
    int j = 0;
    int c;
    int keyk[256];
    for (int i = 0; i<key.length();i++){
        keyk[i]=int(key[i]);
    }
    for(int i = 0; i<255; i++){
        j = (j + T[i] + keyk[i % key.length()]) % 256;
        c =T[i];
        T[i] = T[j];
        T[j] = c;
    }
    return 0;
}

int PRGA(){ //Pseudo random generation algorithm
    int i = 0;
    int j = 0;
    int c;
    int counter = 0;
    while (counter <= text.length()){   // Folosim lungimea textului de ecriptata ca sa stim cat stream trebuie sa generam
        i = (i+1)%256;
        j = (j+T[i])%256;
        c =T[i];
        T[i] = T[j];
        T[j] = c;
        S[counter]=T[(T[i] + T[j]) % 256];
        counter++;
    }
    return 0;
}
int main(){
    KSA();
    PRGA();
    for(int i=0;i<text.length();i++){
        enc[i]=char(int(text[i])^S[i]);
            cout<<char(int(text[i])^S[i]);  //XOR
    }
    cout<<endl;

}
